server/src/main/java/org/elasticsearch/cluster/metadata/MetadataIndexTemplateService.java

```diff
@@ -663,6 +663,27 @@ public static String findV2Template(Metadata metadata, String indexName, @Nullab
             return null;
         }
 
+        // this is complex but if the index is not hidden in the create request but is hidden as the result of template application,
+        // then we need to exclude global templates
+        if (isHidden == null) {
+            final Optional<Map.Entry<IndexTemplateV2, String>> templateWithHiddenSetting = matchedTemplates.entrySet().stream()
+                .filter(entry -> IndexMetadata.INDEX_HIDDEN_SETTING.exists(resolveSettings(metadata, entry.getValue()))).findFirst();
+            if (templateWithHiddenSetting.isPresent()) {
+                Map.Entry<IndexTemplateV2, String> indexTemplateV2ToNameEntry = templateWithHiddenSetting.get();
+                final boolean templateIsHidden =
+                    IndexMetadata.INDEX_HIDDEN_SETTING.get(resolveSettings(metadata, indexTemplateV2ToNameEntry.getValue()));
+                if (templateIsHidden) {
+                    // remove the global templates
+                    matchedTemplates.keySet().removeIf(current -> current.indexPatterns().stream().anyMatch(Regex::isMatchAllPattern));
+                }
+                // validate that hidden didn't change
+                if (matchedTemplates.containsKey(indexTemplateV2ToNameEntry.getKey()) == false) {
+                    throw new IllegalStateException("A global index V2 template [" + indexTemplateV2ToNameEntry.getValue() +
+                        "] defined the index hidden setting, which is not allowed");
```

Comment:
https://github.com/elastic/elasticsearch/pull/55010 guards against this when creating a v2 template